rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -499,24 +502,30 @@ impl CanisterQueues {
             let input_queue = &mut self.canister_queues.get_mut(&sender).unwrap().0;
             if input_queue.len() != 0 {
                 input_schedule.push_back(sender);
+            } else {
+                assert!(self.input_schedule_canisters.remove(&sender));
             }
         }
+        debug_assert_eq!(Ok(()), self.test_invariants());
     }
 
     /// Returns `true` if `ingress_queue` or at least one of the canister input
-    /// queues is not empty; `false` otherwise.
+    /// queues contains non-stale messages; `false` otherwise.
     pub fn has_input(&self) -> bool {
-        !self.ingress_queue.is_empty() || self.input_queues_stats.message_count > 0
+        !self.ingress_queue.is_empty() || self.pool.message_stats().inbound_message_count > 0
     }
 
-    /// Returns `true` if at least one output queue is not empty; false
-    /// otherwise.
+    /// Returns `true` if at least one output queue contains non-stale messages;
+    /// false otherwise.
     pub fn has_output(&self) -> bool {
-        self.output_queues_stats.message_count > 0
+        self.pool.message_stats().outbound_message_count > 0
     }
 
-    /// Peeks the next inter-canister or ingress message (round-robin) from
-    /// `self.subnet_queues`.
+    /// Peeks the ingress or inter-canister input message that would be returned by
+    /// `pop_input()`.
+    ///
+    /// Requires a `&mut self` reference in order to be able to drop empty queues
+    /// from the input schedule, in order to achieve amortized `O(1)` time complexity.
     pub(crate) fn peek_input(&mut self) -> Option<CanisterMessage> {
```

Comment:
Stupid question (not really related to your PR): if `peek_input` updates `next_input_queue`, how is it guaranteed that the next `pop_input` will pop the same message as peek did?

Reply:
Oh, nevermind. I missed that `peek_input` only updates `next_input_queue` if the current one is empty.